### PR TITLE
NEW: Query Cache / Eager loading

### PR DESCRIPTION
**Pull request by unclecheese**
Issue https://github.com/silverstripe/silverstripe-framework/issues/8426

### CODE REVIEW

**Comment by unclecheese** [comment]
Worth looking into -- should we stale the query cache when an object is written during the same request cycle? How aggressive should that be? Don't forget to account for inheritance, e.g. a write to Page needs to stale ProductPage, and vice-versa (I think.... )

**Comment by patricknelson** [comment]
Here's a thought: Is there any major _disadvantage_ to staling the entire query cache on any write operation (update', 'delete', 'replace', 'drop', 'create', 'truncate' and 'alter`)? Off-hand I can think of one minor one, which is if you're using DB driven sessions (like I am) then you have a write but that's only in the shutdown function.

The advantage of eager loading is primarily when fetching objects for front-end display and a vast majority of operations are going to be read-only. I can see where the framework is being used in write-heavy situations (e.g. maybe a forum) in a front-end context. I feel like you could setup a whitelist or blacklist where you would identify the objects where, if modified in any way, will (or will not) purge the cache for those edge cases where a developer will need to eek out a bit more performance.

**Comment by Cheddam** [comment]
@unclecheese Is there a clear path to shipping this? Would be great to see it in 4.7 (which we'll be looking at releasing in December at this stage.)

**Comment by unclecheese** [comment]
Yup, should be a very clear path. All the tests were passing when I published the PR, and I've resolved all the conflicts. What it really needs is a formal peer review, check for performance regressions, etc., and documentation. I'm happy to start on the latter in parallel.

**Comment by sminnee** [comment]
It would be useful to pull this in with composer-patches to a real-world project or two and see if it causes issues beyond the test failures.

A couple of things for that to happen:

 * Tests should pass
 * It need to be rebased, rather than the 4 branch merged into the PR branch

**Comment by unclecheese** [comment]
Another thing that shouldn't be overlooked is that this is key to getting the ElementalEditor field faster. As we found with TNZ, it's riddled with N+1s, sometimes three levels deep.

File: src/ORM/DataQuery.php [hunk 1]
```diff
@@ -21,6 +22,15 @@
  */
 class DataQuery
```
Comment (sminnee) [inline src/ORM/DataQuery.php:23]: This refactor has the biggest chance of regressions. Anything that subclassed DataQuery would be in trouble, but that's rare and unrecommended, AFAIK.

It would be good to test the modification on sites that make a lot of use of constructing their own DataQueries, though.

File: src/ORM/DataQueryExecutorInterface.php [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+use Iterator;
+
+interface DataQueryExecutorInterface
+{
+    /**
+     * @param DataQuery $dataQuery
+     * @param string $modifier
+     * @return iterable
+     */
+    public function execute(DataQuery $dataQuery, ?string $modifier = null): iterable;
```
Comment (sminnee) [inline src/ORM/DataQueryExecutorInterface.php:14]: Add an additional argument to all of these `FutureQueryHints $hints`.

File: src/ORM/EagerLoading/DataListEagerLoader.php [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+namespace SilverStripe\ORM\EagerLoading;
+
+use SilverStripe\ORM\DataList;
+use SilverStripe\ORM\QueryCache\DataQueryStoreInterface;
+use SilverStripe\ORM\DataObject;
+use InvalidArgumentException;
+use SilverStripe\Core\Injector\Injector;
+use Psr\Container\NotFoundExceptionInterface;
+
+class DataListEagerLoader implements QueryEagerLoaderInterface
```
Comment (sminnee) [inline src/ORM/EagerLoading/DataListEagerLoader.php:12]: This would implement DataQueryExecutorInterface instead. It would take a cached query executor as a property, and pass all calls to that. performEagerLoading would be called from within its implementations of execute, getFirstRow, getLastRow, based on the eager relations specified in the FutureQueryHints object passed as an argument to those methods, rather than $this->eagerLoad.

**Review by sminnee — COMMENTED** [review]
In general this looks great, and while I think we need to test this code against some projects that make direct use of the DataQuery class, we should totally get this into 4.x.

I've suggested a refactor, that pushes eager loading to being a special kind of DataQueryExecutorInterface, rather than a 2nd per-instance object that DataList also manages.

This simplifies the coordination that DataList needs to do, but introduces an additional value object, FutureQueryHints, that DataList would need to construct based on calls to DataList::with(), and pass to the DataQueryExecutor along with the DataQuery.

Happy to be challenged on that refactor, but I think it decouples eager loading from DataList in a valuable way.

File: src/ORM/EagerLoading/BelongsToEagerLoader.php [hunk 4]
```diff
@@ -0,0 +1,8 @@
+<?php
+
+namespace SilverStripe\ORM\EagerLoading;
+
+class BelongsToEagerLoader
```
[the remark this answers is not in the document]
Comment (unclecheese) [reply src/ORM/EagerLoading/BelongsToEagerLoader.php:5]: I think it is, yes.. Who uses `belongs_to`? 😂

[the remark this answers is not in the document]
Comment (sminnee) [reply src/ORM/EagerLoading/BelongsToEagerLoader.php:5]: It belongs to no-one.

File: src/ORM/DataQueryExecutorInterface.php [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+
+    /**
+     * @param DataQuery $dataQuery
+     * @return iterable|null
+     */
+    public function getLastRow(DataQuery $dataQuery): ?iterable;
+
+    /**
+     * @param DataQuery $dataQuery
+     * @return string
```
[the remark this answers is not in the document]
Comment (unclecheese) [reply src/ORM/DataQueryExecutorInterface.php:30]: I'm pretty sure `DataList::count()` does return a string, though. Otherwise, I'm sure this would have thrown in tests.

**Comment by sminnee** [comment]
Is this ready for review now Aaron? 😍

File: src/ORM/DataQueryExecutorInterface.php [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+
+namespace SilverStripe\ORM;
+
+use Iterator;
+
+interface DataQueryExecutorInterface
+{
+    /**
+     * @param DataQuery $dataQuery
+     * @param string $modifier
```
[the remark this answers is not in the document]
Comment (unclecheese) [reply src/ORM/DataQueryExecutorInterface.php:9]: I don't believe we need this parameter anymore.

**Comment by unclecheese** [comment]
@sminnee Have refactored per your suggestions, and tests are passing again. Getting there. Still need the BelongsTo loader implemented and maybe a bit more test coverage, but have a look.

**Comment by sminnee** [comment]
As a general comment when doing code review: If we're looking to introduce this into a stable branch the most important thing is that the behaviour without with() queries is not put at risk. As people start to use with() in production, edge-case issues will undoubtedly be found and we can address them; but we don't want to risk breaking existing sites.

Avoiding the use of the new query executor without explicit config is one approach to this; however we'd also want to avoid risks to sites by switching to the new executor but never using with() - otherwise enabling the new executor will become a very risky operation.

In other words, with the new executor enabled and with() used on a single ORM query, I want to feel very safe for all the other queries in the system.

From my perspective, that's the main thing blocking it from a stable release.

I'd also encourage us to make use of this on a few projects running 4.x-dev prior to a stable release. (RMS is probably a good internal project candidate for that.)

**Comment by unclecheese** [comment]
Yeah you can see heaps of tests failing right now because the cached executor is enabled in the PR, so I think we should prove that everything is green with the naive implementation, and if the cached implementation breaks a few tests when enabled, we can document as unstable.

**Comment by sminnee** [comment]
We may wish to consider a build matrix that runs the relevant tests (maybe just the ORM ones?) with the separate executor on as well as off.

**Comment by obj63mc** [comment]
With SS5 coming up in the near future, wondering if this has any chance of being finished and merged?  One of the main issues we have with SS on large sites is this n+1 issue.

**Comment by emteknetnz** [comment]
Closing in favor of https://github.com/silverstripe/silverstripe-framework/pull/10783
